File: Python/day25/AOC.py
from itertools import product

def parse_input(input):
    locks, keys = [], []
    for s in input.split("\n\n"):
        if s[0] == '.':
            transposed = [list(x) for x in zip(*s.split("\n"))] # learned this last year, tanspose rows and columns
            locks.append([6 - ''.join(line).index('#') for line in transposed]) # 6 is the height of the lock/key
        else:
            transposed = [list(x) for x in zip(*s.split("\n"))]
            keys.append([''.join(line).rindex('#') for line in transposed])

    return locks, keys

def part1(input):
    return sum(1 for key, lock in product(*parse_input(input)) if all(k + l < 6 for k, l in zip(key, lock)))

def main():
    input = open("input.txt", "r").read()
    print(f"Part 1: {part1(input)}")

if __name__ == "__main__":
    main()

File: Go/day25/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
	"utils"
)

func parseInput(input string) ([][]int, [][]int) {
	locks := [][]int{}
	keys := [][]int{}
	for _, line := range strings.Split(input, "\n\n") {
		count := utils.Repeat(-1, 5)
		for _, row := range strings.Split(line, "\n") {
			for j, char := range row {
				if char == '#' {
					count[j]++
				}
			}
		}
		if line[0] == '#' {
			locks = append(locks, count)
		} else {
			keys = append(keys, count)
		}
	}

	return locks, keys
}

func part1(input string) int {
	locks, keys := parseInput(input)
	fit := func(lock, key []int) bool {
		for i := range lock {
			if lock[i]+key[i] > 5 {
				return false
			}
		}
		return true
	}

	return utils.ApplySum(locks, func(lock []int) int {
		return utils.ApplySum(keys, func(key []int) int {
			if fit(lock, key) {
				return 1
			}
			return 0
		})
	})
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := string(input)
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)
}
